fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/ProgressiveLoanInterestRefundServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+public class ProgressiveLoanInterestRefundServiceImpl implements InterestRefundService {
+
+    private final AdvancedPaymentScheduleTransactionProcessor processor;
+    private final EMICalculator emiCalculator;
+
+    private static LoanRepaymentScheduleInstallment duplicate(LoanRepaymentScheduleInstallment rs) {
+        MonetaryCurrency currency = rs.getLoan().getCurrency();
+
+        return new LoanRepaymentScheduleInstallment(rs.getLoan(), rs.getInstallmentNumber(), rs.getFromDate(), rs.getDueDate(),
+                rs.getPrincipal(currency).getAmount(), rs.getInterestCharged(currency).getAmount(),
+                rs.getFeeChargesCharged(currency).getAmount(), rs.getPenaltyChargesCharged(currency).getAmount(),
+                rs.isRecalculatedInterestComponent(), rs.getLoanCompoundingDetails(), rs.getRescheduleInterestPortion(),
+                rs.isDownPayment());
+    }
+
+    private ProgressiveLoanInterestScheduleModel getModelSafely(final Loan loan, BigDecimal refundAmount) {
+        final AtomicReference<BigDecimal> refundFinal = new AtomicReference<>(refundAmount);
+        List<LoanTransaction> withoutSpecificDisbursements = loan.getLoanTransactions().stream()
+                .filter(lt -> !lt.isAccrual() && !lt.isAccrualActivity() && !lt.isReversed()).filter(lt -> {
+                    if (lt.getTypeOf().isDisbursement() && refundFinal.get().compareTo(BigDecimal.ZERO) > 0) {
+                        //
+                        if (lt.getAmount().compareTo(refundFinal.get()) <= 0) {
+                            refundFinal.set(refundFinal.get().subtract(lt.getAmount()));
+                            return false;
+                        }
+                    }
+                    return true;
+                }).toList();
+        List<LoanTransaction> transactionsToReprocess = withoutSpecificDisbursements.stream().map(lt -> {
+            if (lt.getTypeOf().isDisbursement() && refundFinal.get().compareTo(BigDecimal.ZERO) > 0) {
+                refundFinal.set(BigDecimal.ZERO);
+                return new LoanTransaction(lt.getLoan(), lt.getLoan().getOffice(), lt.getTypeOf().getValue(), lt.getDateOf(),
+                        lt.getAmount().subtract(refundFinal.get()), lt.getPrincipalPortion(), lt.getInterestPortion(),
+                        lt.getFeeChargesPortion(), lt.getPenaltyChargesPortion(),
+                        lt.getOverPaymentPortion(lt.getLoan().getCurrency()).getAmount(), lt.isReversed(), lt.getPaymentDetail(),
+                        lt.getExternalId());
+            }
+            return new LoanTransaction(lt.getLoan(), lt.getLoan().getOffice(), lt.getTypeOf().getValue(), lt.getDateOf(), lt.getAmount(),
+                    lt.getPrincipalPortion(), lt.getInterestPortion(), lt.getFeeChargesPortion(), lt.getPenaltyChargesPortion(),
+                    lt.getOverPaymentPortion(lt.getLoan().getCurrency()).getAmount(), lt.isReversed(), lt.getPaymentDetail(),
+                    lt.getExternalId());
```

Comment:
Since you've already created a copy method for the LoanRepaymentScheduleInstallment, can't we create a similar one for LoanTransaction as well? Parameter usage seems excessive especially we can easily miss a param.

Reply:
The code was not forward compatible, so I reworked calculation and related mechanics.